posts/2021-04-30-edition-2021.md

```diff
@@ -0,0 +1,346 @@
+This *only* applies to the `.into_iter()` method call syntax.
+It does not affect any other syntax such as `for e in [1, 2, 3]` or `iter.zip([1, 2, 3])`.
+Those will start to work in *all* editions.
+
+While it's a shame that this required a small hack to avoid breakage,
+we're very happy with how this solution keeps the difference between
+the editions to an absolute minimum.
+
+[25]: https://github.com/rust-lang/rust/issues/25725
+[20]: https://github.com/rust-lang/rust/pull/65819
+[22]: https://doc.rust-lang.org/book/ch05-03-method-syntax.html#wheres-the---operator
+
+### Disjoint capture in closures
+
+[Closures](https://doc.rust-lang.org/book/ch13-01-closures.html)
+automatically capture anything that you refer to from within their body.
+For example, `|| a + 1` automatically captures a reference to `a` from the surrounding context.
+
+Currently, this applies to whole structs, even when only using one field.
+For example, `|| a.x + 1` captures a reference to `a` and not just `a.x`.
+In some situations, this is a problem.
+When a field of the struct is already borrowed (mutably) or moved out of,
+the other fields can no longer be used in a closure,
+since that would capture the whole struct, which is no longer available.
+
+```rust
+let a = SomeStruct::new();
+
+drop(a.x); // Move out of one field of the struct
+
+println!("{}", a.y); // Ok: Still use another field of the struct
+
+let c = || println!("{}", a.y); // Error: Tries to capture all of `a`
+c();
+```
+
+Starting in Rust 2021, closures will only capture the fields that they use.
+So, the above example will compile fine in Rust 2021.
+
+This new behavior is only activated in the new edition,
+since it can change the order in which fields are dropped.
+As for all edition changes, an automatic migration is available,
+which will update your closures for which this matters.
+It can insert `let _ = &a;` inside the closure to force the entire
+struct to be captured as before.
+
+### Panic macro consistency
+
+The `panic!()` macro is one of Rust's most well known macros.
+However, it has [some subtle surprises](https://github.com/rust-lang/rfcs/blob/master/text/3007-panic-plan.md)
+that we can't just change due to backwards compatibility.
+
+```rust
+panic!("{}", 1); // Ok, panics with the message "1"
+panic!("{}"); // Ok, panics with the message "{}"
+```
+
+The `panic!()` macro only uses string formatting when it's invoked with more than one argument.
+When invoked with a single argument, it doesn't even look at that argument.
+
+```rust
+let a = "{";
+println!(a); // Error: First argument must be a format string literal
+panic!(a); // Ok: The panic macro doesn't care
+```
+
+(It even accepts non-strings such as `panic!(123)`, which is uncommon and rarely useful.)
+
+This is especially a problem when
+[implicit format arguments](https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html)
+will be stabilized.
```

Comment:
```suggestion
This will especially be a problem once
[implicit format arguments](https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html)
stabilize.
```